Refactor!

mistergone opened the pull request:
This commit is a refactor to graph-view.js to make the code cleaner and more readable. This involved moving translations out of the code and into an object so that gettext() isn't called every time the user changes the graph. Instead, it simply populates an object and then uses the object's values. I also
reorganized setTextByAge() so that it handles each possibility in an if statement, which ought to make it easier to track down bugs and make changes.

This also should fix a bug where the incorrect age was shown in the benefits text.

## Changes
- Many changes to `graph-view.js`

## Testing
- Pull it down, start it up, and see what happens!

## Review
- @marteki and @niqjohnson and @higs4281 - Please, for the love of everything, double-check any weird case you can think of. I moved a lot of code around, and I'm afraid I broke stuff.

## Checklist
* [x] Changes are limited to a single goal (no scope creep)
* [x] Code can be automatically merged (no conflicts)
* [x] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [x] Passes all existing automated tests
* [x] New functions include new tests
* [x] New functions are documented (with a description, list of inputs, and expected output)
* [x] Placeholder code is flagged
* [x] Visually tested in supported browsers and devices 
* [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

## Animated GIF
![moss-fire](https://cloud.githubusercontent.com/assets/1490703/15160792/6ae8349c-16c9-11e6-84c9-22fa15af17dc.gif)

marteki commented:
Obligatory "It would be a *lot* easier to test our weird cases and massive code changes if we had browser tests" remark. 😞 

This is gonna take some time to test. I'll look at it today, and tomorrow (and maybe Thursday).

imuchnik commented:
Obligatory "unit tests are tests too" remark. They also make a lot easier to test stuff.

marteki commented:
Spanish translations on the graph for the "Claiming at" line are broken for every year except FRA, for all birth dates tested.

![screen shot 2016-05-10 at 5 55 39 pm](https://cloud.githubusercontent.com/assets/1183435/15163801/b1ae044c-16d8-11e6-83c7-a4061337b0e6.png)

niqjohnson commented:
I wasn't able to reproduce the "claiming at" translation error that @marteki saw. I did notice a few other translations from #177 that aren't working quite right.

DOB | Selected claiming age | Current Spanish message | Correct Spanish message
--- | --------------------- | ----------------------- | -----------------------
7/7/1977 | 67 | 67 de edad es su plena edad de jubilación. | 67 años de edad es su plena edad de jubilación.
7/7/1955 | 66 | 66 and 2 months de edad es su plena edad de jubilación. | 66 años y 2 meses de edad es su plena edad de jubilación.
1/1/1946 | 70 | 70 es la edad máxima para solicitar. | 70 años es la edad máxima para solicitar.

I'll keep poking around to see if any other edge cases are causing trouble.

mistergone commented:
I also don't see the "Claiming at" bug that @marteki sees. @niqjohnson - I'm checking into those errors.

mistergone commented:
@niqjohnson - I've fixed those missing _años y meses_. 🔧 🔧 🐙

marteki commented:
And I'm not looking seeing the "Claiming at" bug anymore, so either the `gettext` change fixed it or (more likely) my box didn't have completely updated files.

niqjohnson commented:
The "claiming at" Spanish messages all look right to me now, thanks!

I also just noticed that the monthly/annual toggle isn't toggling.

mistergone commented:
@niqjohnson - Good catch! Actually, the function that was supposed to handle that, `updateMonthlyAnnual()`, doesn't do anything, and apparently hasn't for a while. I removed it and fixed the place where these values should've been multiplied by 12. 🔥 🚒

marteki commented:
Going through the test cases from Bill's spreadsheet:

| DOB | Age | Monthly benefit | Lifetime value |
|--------|-----------------------|---------------------------------|------------------------------|
| 2/1/1970 | 62 | ✅  | ✅  |
| 2/3/1970 | 62 | ✅  | ✅  |
| 10/3/1953 | 62 | ❌ (should be $935) | ❌  (should be $252,450) |
| 3/3/1955 | 66 | ✅  | ✅  |
| 3/3/1956 | 66 | ✅  | ✅  |
| 3/3/1957 | 66 | ✅  | ✅  |
| 3/3/1958 | 66 | ✅  | ✅  |
| 3/3/1959 | 66 | ✅  | ✅  |
| 4/1/1955 | 66 | ✅  | ✅  |
| 4/1/1956 | 66 | ✅  | ✅  |
| 4/1/1957 | 66 | ✅  | ✅  |
| 4/1/1958 | 66 | ✅  | ✅  |
| 4/1/1959 | 66 | ✅  | ✅  |

mistergone commented:
The 10/3/1953 date gives the same result in `master` or `refactor`, so I think that's not related to changes I made. @higs4281 ?

niqjohnson commented:
👍 

Everything I tested looks right to me now. Beyond everything generally working as expected, the Spanish translations from #177 all look good, and the monthly/annual toggle is toggling without any errors for all the ages I checked.

Nice!

marteki commented:
> The 10/3/1953 date gives the same result in master or refactor, so I think that's not related to changes I made.

Excellent point. We can figure out what's going on there in another PR, if needed.

:+1:

marteki commented:
And from Hector:

> the 940 is correct
because this person is now 66 and 7 months
Lifetime is also correct at 252,850

So the spreadsheet was just outdated for that birthdate. 😒  Consider those as ✅ s.

mistergone commented:
Great @marteki - thanks for checking into that!

Merging this!
